### ISSUE TITLE
env: /usr/local/etc/rc.d/samplevm: No such file or directory

### ISSUE THREAD

**Issue by bridgedata** [Beckhoff/TCBSD_Hypervisor_Samples#1]
While trying to start virtual machine as a service based on the chapter "9.4 Autostart shell scripts" from https://www.beckhoff.com/en-us/download/558086177, I am keep getting "env: /usr/local/etc/rc.d/samplevm: No such file or directory" error despite that all files and permissions look correct. bhyve works fine but not as a script. [TCBSD_Hypervisor_Samples](https://github.com/Beckhoff/TCBSD_Hypervisor_Samples) repos are up to date.

**Comment by dbateamz** [comment]
found the solution it was my network configuration

**Comment by runtimevic** [comment]
> found the solution it was my network configuration

Hello @dbateamz , 
What virtual operating system have you installed? What guide have you followed to get it?

**Comment by bridgedata** [comment]
Hey! I installed Ubuntu 20.04 and followed [TwinCAT_BSD_en.pdf doc ](https://www.beckhoff.com/en-us/download/558086177)

**Comment by runtimevic** [comment]
> Hey! I installed Ubuntu 20.04 and followed [TwinCAT_BSD_en.pdf doc ](https://www.beckhoff.com/en-us/download/558086177)

Hello @dbateamz , 
And how do you see performance? If I also want to have Docker in Debian... 
Can I install TF2000 HMI in TCBSD? Or do I have to create another virtual machine in Windows?...

**Comment by bridgedata** [comment]
I haven't tested the performance yet but pretty sure some docker images should run without the issue.
You can install both HMI runtimes on BSD. Try: doas pkg search hmi

**Comment by runtimevic** [comment]
> I haven't tested the performance yet but pretty sure some docker images should run without the issue. You can install both HMI runtimes on BSD. Try: doas pkg search hmi

Please, if you don't mind, do you have an email to contact you in case I get stuck in any step....

**Comment by bridgedata** [comment]
I'm on the same boat with you. Just started with BSD... Hopefully, Beckhoff team can provide some support.

**Comment by runtimevic** [comment]
> I'm on the same boat with you. Just started with BSD... Hopefully, Beckhoff team can provide some support.

I leave you my email in case you want to contact me it would be very useful...
email: victorduranm@hotmail.com

**Comment by bridgedata** [comment]
The issue has been resolved. I just needed to remove old configuration files and redo make file.

**Comment by runtimevic** [comment]
@dbateamz ,
> The issue has been resolved. I just needed to remove old configuration files and redo make file.

Can you tell exactly what you deleted?

**Comment by bridgedata** [comment]
Try to rename all "samplevm" with your own name. They try to run doas sh "your name" start. If this works you will be able to run make file and start the script as a service.

**Comment by dbateamz** [comment]
Here is my bhyve config:

`#!/bin/sh
# SPDX-License-Identifier: 0BSD
# Copyright (C) 2022 Beckhoff Automation GmbH & Co. KG

set -e
set -u

usage() {
	cat >&2 << EOF
USAGE: ${0##*/} [COMMAND]
Control the virtual machine environment named '${vm_name}'
COMMANDS:
    start       Start the VM
    stop        Shutdown the running VM
    help        Print this help message
EXAMPLES
# run the virual machine
${0##*/} start
# shutdown the virtual machine
${0##*/} stop
EOF
}

# Cleanup any resources used by the VM after shutdown
cleanup() {
	# do not trap during cleanup procedure
	trap '' EXIT
	set +e
	set +u

	pfctl -a "bhf/bhyve/${vm_name}" -f /dev/null > /dev/null 2>&1
	if test -e "${temp_firewall_rules}"; then
		rm "${temp_firewall_rules}"
	fi
}

prepare_host() {
	# Ensure that kernel modul vmm.ko is loaded
	kldload -n vmm.ko

	# accept incoming VNC connections
	temp_firewall_rules="/etc/pf.conf.d/${vm_name}"
	printf "pass in quick proto tcp to port %s\n" "${vm_vnc_port}" > "${temp_firewall_rules}"
	pfctl -a bhf/bhyve/"${vm_name}" -f "${temp_firewall_rules}"
}

# Setups the VM environment and actually executes the bhyve process
run_vm() {
	trap 'cleanup' EXIT

	prepare_host

	while true; do

		if test -e "/dev/vmm/${vm_name}"; then
			bhyvectl --vm="${vm_name}" --destroy
		fi

		_bhyve_rc=0
		bhyve \
			-c sockets=1,cores=2,threads=2 \
			-m 4G \
			-l bootrom,/usr/local/share/uefi-firmware/BHYVE_BHF_UEFI.fd,/vms/samplevm/EFI_VARS.fd \
			-s 0,hostbridge \
			-s 2,fbuf,rfb=0.0.0.0:"${vm_vnc_port}",w=1280,h=1024 \
			-s 3,xhci,tablet \
			-s 10,nvme,/dev/zvol/zroot/vms/samplevm/disk0 \
			-s 20,virtio-net,tap0 \
			-s 31,lpc \
			-A -H -P -w \
			"${vm_name}" || _bhyve_rc=$?

# Ubuntu Installer (add after -s10)
# -s 15,ahci-cd,/vms/ubuntu-installer.iso,ro \

		if test "${_bhyve_rc}" -ne 0; then
			break
		fi

	done
}

shutdown_vm() {
	# do not trap on exit during shutdown command
	trap '' EXIT
	set +e
	set +u

	kill "${_pid}"
}

get_bhyve_pid() {
	printf "%s" "$(pgrep -f "bhyve: ${vm_name}")"
}

# Execution of virtual machines requires root previleges
if test "$(id -u)" -ne 0; then
	printf "%s must be run as root\n" "${0##*/}"
	exit 1
fi

# Default values for VM configuration
readonly vm_name="vmubuntu"
readonly vm_vnc_port="5900"
readonly _cmd="${1?Error: No COMMAND specified$(usage)}"

_pid="$(get_bhyve_pid)"
case "${_cmd}" in
	start)
		if test -z "${_pid:+x}"; then
			run_vm
		else
			printf "%s is already running with pid: %s\n" "${vm_name}" "${_pid}"
			exit 1
		fi
		;;

	stop)
		if ! test -z "${_pid:+x}"; then
			shutdown_vm
		fi
		;;

	status)
		if ! test -z "${_pid:+x}"; then
			printf "'%s' is running with pid: %s\n" "${vm_name}" "${_pid}"
		else
			printf "'%s' is not running.\n" "${vm_name}"
		fi
		;;

	-h | --help | help)
		usage
		;;

	*)
		usage
		printf "Unknown COMMAND: '%s'\n" "${_cmd}"
		exit 1
		;;
esac`

**Comment by bridgedata** [comment]
Here is an example script:
```
#!/bin/sh
# SPDX-License-Identifier: 0BSD
# Copyright (C) 2022 Beckhoff Automation GmbH & Co. KG

set -e
set -u

usage() {
	cat >&2 << EOF
USAGE: ${0##*/} [COMMAND]
Control the virtual machine environment named '${vm_name}'
COMMANDS:
    start       Start the VM
    stop        Shutdown the running VM
    help        Print this help message
EXAMPLES
# run the virual machine
${0##*/} start
# shutdown the virtual machine
${0##*/} stop
EOF
}

# Cleanup any resources used by the VM after shutdown
cleanup() {
	# do not trap during cleanup procedure
	trap '' EXIT
	set +e
	set +u

	pfctl -a "bhf/bhyve/${vm_name}" -f /dev/null > /dev/null 2>&1
	if test -e "${temp_firewall_rules}"; then
		rm "${temp_firewall_rules}"
	fi
}

prepare_host() {
	# Ensure that kernel modul vmm.ko is loaded
	kldload -n vmm.ko

	# accept incoming VNC connections
	temp_firewall_rules="/etc/pf.conf.d/${vm_name}"
	printf "pass in quick proto tcp to port %s\n" "${vm_vnc_port}" > "${temp_firewall_rules}"
	pfctl -a bhf/bhyve/"${vm_name}" -f "${temp_firewall_rules}"
}

# Setups the VM environment and actually executes the bhyve process
run_vm() {
	trap 'cleanup' EXIT

	prepare_host

	while true; do

		if test -e "/dev/vmm/${vm_name}"; then
			bhyvectl --vm="${vm_name}" --destroy
		fi

		_bhyve_rc=0
		bhyve \
			-c sockets=1,cores=2,threads=2 \
			-m 4G \
			-l bootrom,/usr/local/share/uefi-firmware/BHYVE_BHF_UEFI.fd,/vms/samplevm/EFI_VARS.fd \
			-s 0,hostbridge \
			-s 2,fbuf,rfb=0.0.0.0:"${vm_vnc_port}",w=1280,h=1024 \
			-s 3,xhci,tablet \
			-s 10,nvme,/dev/zvol/zroot/vms/samplevm/disk0 \
			-s 20,virtio-net,tap0 \
			-s 31,lpc \
			-A -H -P -w \
			"${vm_name}" || _bhyve_rc=$?

# Ubuntu Installer (add after -s10)
# -s 15,ahci-cd,/vms/ubuntu-installer.iso,ro \

		if test "${_bhyve_rc}" -ne 0; then
			break
		fi

	done
}

shutdown_vm() {
	# do not trap on exit during shutdown command
	trap '' EXIT
	set +e
	set +u

	kill "${_pid}"
}

get_bhyve_pid() {
	printf "%s" "$(pgrep -f "bhyve: ${vm_name}")"
}

# Execution of virtual machines requires root previleges
if test "$(id -u)" -ne 0; then
	printf "%s must be run as root\n" "${0##*/}"
	exit 1
fi

# Default values for VM configuration
readonly vm_name="vmubuntu"
readonly vm_vnc_port="5900"
readonly _cmd="${1?Error: No COMMAND specified$(usage)}"

_pid="$(get_bhyve_pid)"
case "${_cmd}" in
	start)
		if test -z "${_pid:+x}"; then
			run_vm
		else
			printf "%s is already running with pid: %s\n" "${vm_name}" "${_pid}"
			exit 1
		fi
		;;

	stop)
		if ! test -z "${_pid:+x}"; then
			shutdown_vm
		fi
		;;

	status)
		if ! test -z "${_pid:+x}"; then
			printf "'%s' is running with pid: %s\n" "${vm_name}" "${_pid}"
		else
			printf "'%s' is not running.\n" "${vm_name}"
		fi
		;;

	-h | --help | help)
		usage
		;;

	*)
		usage
		printf "Unknown COMMAND: '%s'\n" "${_cmd}"
		exit 1
		;;
esac
```

**Comment by runtimevic** [comment]
Here is my bhyve config:

```
#!/bin/sh
# SPDX-License-Identifier: 0BSD
# Copyright (C) 2022 Beckhoff Automation GmbH & Co. KG

set -e
set -u

usage() {
	cat >&2 << EOF
USAGE: ${0##*/} [COMMAND]

Control the virtual machine environment named '${vm_name}'

COMMANDS:
    start       Start the VM
    stop        Shutdown the running VM
    help        Print this help message

EXAMPLES
# run the virual machine
${0##*/} start

# shutdown the virtual machine
${0##*/} stop

EOF
}

# Cleanup any resources used by the VM after shutdown
cleanup() {
	# do not trap during cleanup procedure
	trap '' EXIT
	set +e
	set +u

	pfctl -a "bhf/bhyve/${vm_name}" -f /dev/null > /dev/null 2>&1
	if test -e "${temp_firewall_rules}"; then
		rm "${temp_firewall_rules}"
	fi
}

prepare_host() {
	# Ensure that kernel modul vmm.ko is loaded
	kldload -n vmm.ko

	# accept incoming VNC connections
	temp_firewall_rules="/etc/pf.conf.d/${vm_name}"
	printf "pass in quick proto tcp to port %s\n" "${vm_vnc_port}" > "${temp_firewall_rules}"
	pfctl -a bhf/bhyve/"${vm_name}" -f "${temp_firewall_rules}"
}

# Setups the VM environment and actually executes the bhyve process
run_vm() {
	trap 'cleanup' EXIT

	prepare_host

	while true; do

		if test -e "/dev/vmm/${vm_name}"; then
			bhyvectl --vm="${vm_name}" --destroy
		fi

		_bhyve_rc=0
		bhyve \
			-A -H -P -w \
			-c sockets=1,cores=1,threads=1 \
			-m 2G \
			-l bootrom,/usr/local/share/uefi-firmware/BHYVE_BHF_UEFI.fd,/vms/samplevm/EFI_VARS.fd \
			-s 0,hostbridge \
			-s 2,fbuf,rfb=0.0.0.0:"${vm_vnc_port}",w=1280,h=1024 \
			-s 3,xhci,tablet \
			-s 10,nvme,/dev/zvol/zroot/vms/samplevm/disk0 \
			-s 15,ahci-cd,/vms/debian-installer.iso,ro \
			-s 20,virtio-net,tap0 \
			-s 31,lpc \
			"${vm_name}" || _bhyve_rc=$?

		if test "${_bhyve_rc}" -ne 0; then
			break
		fi

	done
}

shutdown_vm() {
	# do not trap on exit during shutdown command
	trap '' EXIT
	set +e
	set +u

	kill "${_pid}"
}

get_bhyve_pid() {
	printf "%s" "$(pgrep -f "bhyve: ${vm_name}")"
}

# Execution of virtual machines requires root previleges
if test "$(id -u)" -ne 0; then
	printf "%s must be run as root\n" "${0##*/}"
	exit 1
fi

# Default values for VM configuration
readonly vm_name="samplevm"
readonly vm_vnc_port="5900"
readonly _cmd="${1?Error: No COMMAND specified$(usage)}"

_pid="$(get_bhyve_pid)"
case "${_cmd}" in
	start)
		if test -z "${_pid:+x}"; then
			run_vm
		else
			printf "%s is already running with pid: %s\n" "${vm_name}" "${_pid}"
			exit 1
		fi
		;;

	stop)
		if ! test -z "${_pid:+x}"; then
			shutdown_vm
		fi
		;;

	status)
		if ! test -z "${_pid:+x}"; then
			printf "'%s' is running with pid: %s\n" "${vm_name}" "${_pid}"
		else
			printf "'%s' is not running.\n" "${vm_name}"
		fi
		;;

	-h | --help | help)
		usage
		;;

	*)
		usage
		printf "Unknown COMMAND: '%s'\n" "${_cmd}"
		exit 1
		;;
esac

```

I am using the manual: TwinCAT/BSD Operating system 12/12/2022 version:1.3, on page 65, I go to point 7:
doas sh samplevm start
I attached a photo of the error it gives me...
how can I solve it?

<img width="739" alt="Beckhoff_TCBSD_Hypervisor_install_error_01" src="https://user-images.githubusercontent.com/53791834/218500608-895c4a2e-71b7-4ecf-ab38-9bb023712418.PNG">

<img width="156" alt="Beckhoff_TCBSD_Hypervisor_install_error_02" src="https://user-images.githubusercontent.com/53791834/218500940-7b8247fe-aa2d-4470-9cc2-e9e481f00376.PNG">

**Comment by bridgedata** [comment]
Did you create a new zroot disk, copy EFI_VARS.fd into /vms/samplevm/, and bridge your network interface?

**Comment by runtimevic** [comment]
> Did you create a new zroot disk, copy EFI_VARS.fd into /vms/samplevm/, and bridge your network interface?

<img width="132" alt="Beckhoff_TCBSD_Hypervisor_install_error_03" src="https://user-images.githubusercontent.com/53791834/218501315-8193506d-7d62-44e7-8cb1-2b88632d0b3d.PNG">


<img width="530" alt="Beckhoff_TCBSD_Hypervisor_install_error_04" src="https://user-images.githubusercontent.com/53791834/218502257-ec6bec0a-2111-487e-b9b0-fa38bc3a5f87.PNG">

**Comment by bridgedata** [comment]
make sure your samplevm has sufficient permissions to execute the script as root... it looks you can read but not execute...

**Comment by bridgedata** [comment]
https://docs.freebsd.org/doc/6.3-RELEASE/usr/share/doc/handbook/permissions.html

**Comment by bridgedata** [comment]
chmod +x samplevm

**Comment by runtimevic** [comment]
> https://docs.freebsd.org/doc/6.3-RELEASE/usr/share/doc/handbook/permissions.html

<img width="522" alt="Beckhoff_TCBSD_Hypervisor_install_error_08" src="https://user-images.githubusercontent.com/53791834/218505688-4f239e92-7809-4e4e-910f-d458d62fdd0f.PNG">

**Comment by runtimevic** [comment]
> chmod +x samplevm

sudo chmod 777 /usr/home/

**Comment by bridgedata** [comment]
samplevm script still has -rw-r--r-- which is read only

**Comment by runtimevic** [comment]
> samplevm script still has -rw-r--r-- which is read only

Hello, @dbateamz 
I still can't solve it, do you know what I can do?
<img width="412" alt="Beckhoff_TCBSD_Hypervisor_install_error_09" src="https://user-images.githubusercontent.com/53791834/218673898-acc17873-be7c-4c45-8883-52296baecfb2.PNG">

<img width="443" alt="Beckhoff_TCBSD_Hypervisor_install_error_10" src="https://user-images.githubusercontent.com/53791834/218696260-e82a5f31-d0f1-4dc6-9e21-72846f22d6b4.PNG">

**Comment by bridgedata** [comment]
can you try to run "doas make", "doas service samplevm enable", and then "doas service samplevm start"?

**Comment by bridgedata** [comment]
![MicrosoftTeams-image](https://user-images.githubusercontent.com/110982058/218742340-49fd37e8-a2bb-4006-91e3-717c928a3425.png) This is what I got.

Did you try to contact Beckhoff?

**Comment by runtimevic** [comment]
> doas service samplevm enable

<img width="1008" alt="Beckhoff_TCBSD_Hypervisor_install_error_11" src="https://user-images.githubusercontent.com/53791834/218746791-8df39e5a-82f3-4add-aee5-846734fa3ae6.PNG">

**Comment by runtimevic** [comment]
> ![MicrosoftTeams-imagen](https://user-images.githubusercontent.com/110982058/218742340-49fd37e8-a2bb-4006-91e3-717c928a3425.png)Esto es lo que conseguí.
> 
> ¿Intentaste contactar a Beckhoff?



> ![MicrosoftTeams-image](https://user-images.githubusercontent.com/110982058/218742340-49fd37e8-a2bb-4006-91e3-717c928a3425.png) This is what I got.
> 
> Did you try to contact Beckhoff?

yes, but they couldn't help me...
Installing a guest operating system using Debian Linux as an example:
https://infosys.beckhoff.com/content/1033/twincat_bsd/12607719691.html?id=6880469216400983879

**Comment by runtimevic** [comment]
Hello @dbateamz ,
You have followed chapter:
(9.4-Autostart shell scripts).
I instead am following the chapter:
(9.11-Installing a guest operating system using Debian
Linux as an example).
My question is why have you followed chapter 9.4 instead of 9.11?.If the objective is to install a linux virtual machine, can you post the steps you followed to achieve it and the files? So I can try exactly how you do it, thank you very much.

**Comment by bridgedata** [comment]
This tutorial worked for me, and I won't be able to you help any further.

In the following sample Debian is to be installed in a virtual machine under TwinCAT/BSD. The shell scripts
from the GitHub repository https://github.com/Beckhoff/TCBSD_Hypervisor_Samples/tree/main/vm_autostart
can be used as a template.


```
vm_autostart
├── Makefile
├── rc.d
│ └── vmubuntu
└── vmubuntu
```

The bhyve call implies the following configuration steps:
1. the creation of virtual hard disks
2. the use of ISO images
3. booting a UEFI-based virtual machine
4. VNC based interaction with the virtual machine and
5. the configuration of a bridged network

1. As described in the chapter ZFS data sets as storage location for virtual machines [} 56], a ZFS data set
is first created for the virtual machine.
`doas zfs create -p -o mountpoint=/vms/vmubuntu zroot/vms/vmubuntu`
2. For the installation of the operating system the Debian "network install" CD-ISO should be used. The ISO
file can be downloaded with fetch(8), as described in the chapter Use of installation media (ISO
images) [} 58], and later passed to the bhyve call as a disk of an ahci-hd device:

```
doas fetch -o /vms/debian-installer.iso https://cdimage.debian.org/debian-cd/current/amd64/isocd/
debian-11.5.0-amd64-netinst.iso
```

3. To be able to install Debian on a virtual hard disk, the following command creates a ZFS volume which is
used as backend for the emulated nvme device in the upper bhyve call:
`doas zfs create -V 20G zroot/vms/vmubuntu/disk0`
4. Debian uses EFI variables to store information about bootable disks. Therefore, a copy of the /usr/
local/share/uefi-firmware/BHYVE_BHF_UEFI_VARS.fd file should be placed on the ZFS data
set for the virtual machine:
`doas cp /usr/local/share/uefi-firmware/BHYVE_BHF_UEFI_VARS.fd /vms/vmubuntu/EFI_VARS.fd`
5. For the Debian installation, the virtual machine requires an Internet connection. For this purpose, a
bridged network [} 60] is created on the TwinCAT/BSD host to which the virtual machine is connected
via a virtio-net based network interface and the tap0 instance.
8. In addition, the virtual machine should be able to be operated via a VNC connection in order to be able to
use the graphical installation of the Debian installer at the first start. The vmubuntu shell script therefore
configures the packet filter rules to allow incoming TCP connections on port 5900 of the TwinCAT/BSD
host.
9. With the customized bhyve(8) call within the vmubuntu shell script, the virtual machine can be started
as follows:
`doas sh vmubuntu start` 
or 
`doas service vmubuntu enable`
and 
`doas service vmubuntu start`
10. The started bhyve(8) process then generates the following output on the command line:
fbuf frame buffer base: 0x881e00000 [sz 16777216]
11. Now a VNC client, such as Ultra-VNC can be used to connect to the virtual machine

Once the installation of Debian is complete, the virtual machine is restarted. This will terminate the VNC
connection. After reconnecting, the Debian operating system can be used in the virtual machine.

# **Autostart script named "vmubuntu"**

```
#!/bin/sh
# SPDX-License-Identifier: 0BSD
# Copyright (C) 2022 Beckhoff Automation GmbH & Co. KG

set -e
set -u

usage() {
	cat >&2 << EOF
USAGE: ${0##*/} [COMMAND]
Control the virtual machine environment named '${vm_name}'
COMMANDS:
    start       Start the VM
    stop        Shutdown the running VM
    help        Print this help message
EXAMPLES
# run the virual machine
${0##*/} start
# shutdown the virtual machine
${0##*/} stop
EOF
}

# Cleanup any resources used by the VM after shutdown
cleanup() {
	# do not trap during cleanup procedure
	trap '' EXIT
	set +e
	set +u

	pfctl -a "bhf/bhyve/${vm_name}" -f /dev/null > /dev/null 2>&1
	if test -e "${temp_firewall_rules}"; then
		rm "${temp_firewall_rules}"
	fi
}

prepare_host() {
	# Ensure that kernel modul vmm.ko is loaded
	kldload -n vmm.ko

	# accept incoming VNC connections
	temp_firewall_rules="/etc/pf.conf.d/${vm_name}"
	printf "pass in quick proto tcp to port %s\n" "${vm_vnc_port}" > "${temp_firewall_rules}"
	pfctl -a bhf/bhyve/"${vm_name}" -f "${temp_firewall_rules}"
}

# Setups the VM environment and actually executes the bhyve process
run_vm() {
	trap 'cleanup' EXIT

	prepare_host

	while true; do

		if test -e "/dev/vmm/${vm_name}"; then
			bhyvectl --vm="${vm_name}" --destroy
		fi

		_bhyve_rc=0
		bhyve \
			-c sockets=1,cores=2,threads=2 \
			-m 4G \
			-l bootrom,/usr/local/share/uefi-firmware/BHYVE_BHF_UEFI.fd,/vms/vmubuntu/EFI_VARS.fd \
			-s 0,hostbridge \
			-s 2,fbuf,rfb=0.0.0.0:"${vm_vnc_port}",w=1280,h=1024 \
			-s 3,xhci,tablet \
			-s 10,nvme,/dev/zvol/zroot/vms/vmubuntu/disk0 \
			-s 20,virtio-net,tap0 \
			-s 31,lpc \
			-A -H -P -w \
			"${vm_name}" || _bhyve_rc=$?

# Ubuntu Installer (add after -s10)
# -s 15,ahci-cd,/vms/ubuntu-installer.iso,ro \

		if test "${_bhyve_rc}" -ne 0; then
			break
		fi

	done
}

shutdown_vm() {
	# do not trap on exit during shutdown command
	trap '' EXIT
	set +e
	set +u

	kill "${_pid}"
}

get_bhyve_pid() {
	printf "%s" "$(pgrep -f "bhyve: ${vm_name}")"
}

# Execution of virtual machines requires root previleges
if test "$(id -u)" -ne 0; then
	printf "%s must be run as root\n" "${0##*/}"
	exit 1
fi

# Default values for VM configuration
readonly vm_name="vmubuntu"
readonly vm_vnc_port="5900"
readonly _cmd="${1?Error: No COMMAND specified$(usage)}"

_pid="$(get_bhyve_pid)"
case "${_cmd}" in
	start)
		if test -z "${_pid:+x}"; then
			run_vm
		else
			printf "%s is already running with pid: %s\n" "${vm_name}" "${_pid}"
			exit 1
		fi
		;;

	stop)
		if ! test -z "${_pid:+x}"; then
			shutdown_vm
		fi
		;;

	status)
		if ! test -z "${_pid:+x}"; then
			printf "'%s' is running with pid: %s\n" "${vm_name}" "${_pid}"
		else
			printf "'%s' is not running.\n" "${vm_name}"
		fi
		;;

	-h | --help | help)
		usage
		;;

	*)
		usage
		printf "Unknown COMMAND: '%s'\n" "${_cmd}"
		exit 1
		;;
esac
```


you need also makefile and rc.d file to start the script as system service. see both scripts below:

# **Makefile**
```
PREFIX?=/usr/local
BINDIR=$(DESTDIR)$(PREFIX)/bin
RCDIR=$(DESTDIR)$(PREFIX)/etc/rc.d

INSTALL=/usr/bin/install
MKDIR=/bin/mkdir

SCRIPT=vmubuntu

install:
	$(MKDIR) -p $(BINDIR)
	$(INSTALL) -m 544 $(SCRIPT) $(BINDIR)/

	$(MKDIR) -p $(RCDIR)
	$(INSTALL) -m 555 rc.d/* $(RCDIR)/
```

# **Rc.d/vmubuntu**

```
#!/bin/sh
# SPDX-License-Identifier: 0BSD
# Copyright (C) 2019 - 2022 Beckhoff Automation GmbH & Co. KG

# PROVIDE: vmubuntu
# REQUIRE: DAEMON NETWORKING dmesg
# BEFORE: LOGIN TcSystemService
# KEYWORD: nojail shutdown

. /etc/rc.subr

name="vmubuntu"
rcvar="${name}_enable"

command="/usr/local/bin/${name}"

export PATH="${PATH}:/usr/local/bin:/usr/local/sbin"

start_cmd="${command} start > /dev/null 2>&1 &"
stop_cmd="${command} stop > /dev/null 2>&1 &"
status_cmd="${command} status"

load_rc_config "${name}"
run_rc_command "$1"
```

quick test of running vms:  `doas service vmubuntu status` or `ps -a | grep bhyve`

**Comment by runtimevic** [comment]
A question in which directory do you paste the vm_autostart folder? To put it exactly the same, what do you install Debian or Ubuntu? if you have installed ubuntu what flavor have you tried?

**Comment by bridgedata** [comment]
Local folder (the same as an example). I installed Ubuntu Server 22.04.1 following Beckhoff tutorial and Windows 10 IoT Enterprise following this tutorial https://wiki.freebsd.org/bhyve/Windows. I hope this will help.

**Comment by runtimevic** [comment]
the directory in which I paste the folder is? 
cd /usr/home/vm_autostart

**Comment by bridgedata** [comment]
/usr/home/Administrator/vm_autostart

**Comment by runtimevic** [comment]
- In step 7 of the installation, don't you do it first of all?  : 
doas make

**Comment by runtimevic** [comment]
Hello, @dbateamz 
I finally found the problem: 
https://confluence.atlassian.com/kb/starting-service-on-linux-throws-a-no-such-file-or-directory-error-794203722.html
```
sed -i -e 's/\r//g' /usr/home/Administrator/vm_autostart-Ubuntu/vmubuntu
```

**Comment by runtimevic** [comment]
@dbateamz, sorry...
```
$ doas service vmubuntu start
$ ps -a | grep bhyve
1927  0  SC   0:47.59 bhyve: vmubuntu (bhyve)
1980  0  S+   0:00.00 grep bhyve
$ doas service vmubuntu status
'vmubuntu' is running with pid: 1927
```

- How can I make the start persistent so that the virtual machine starts by itself?
- If I enter I.P::5900 again, I get the steps to reinstall the MV again..., I have already done the steps to install ubuntu server before, do I have to do any final step?
- Do I have to comment on this?, Should I comment it in the file that is in the following path?
/usr/local/bin/ or /home/Administrator/vm_autostart-Ubuntu/
```
# Ubuntu Installer (add after -s10)
# -s 15,ahci-cd,/vms/ubuntu-installer.iso,ro \
```

**Comment by runtimevic** [comment]
> 

Hello, @dbateamz 
I finally solved it with crontab...
Is it the correct solution? Have you also used them? so that the virtual machine always boots...
```
/etc/cron.d
crontab --> @reboot sleep 600 && /usr/local/bin/startubuntuvm.sh
```
file --> startubuntuvm.sh
```
fi
doas bhyve -c sockets=1,cores=2,threads=2 \
-m 4G \
-l bootrom,/usr/local/share/uefi-firmware/BHYVE_BHF_UEFI.fd,/vms/vmubuntu/EFI_VARS.fd \
-s 0,hostbridge \
-s 2,fbuf,rfb=0.0.0.0:5900,w=1280,h=1024 \
-s 3,xhci,tablet \
-s 10,nvme,/dev/zvol/zroot/vms/vmubuntu/disk0 \
-s 20,virtio-net,tap0 \
-s 31,lpc \
-A -H -P -w \
vmubuntu
```

**Comment by runtimevic** [comment]
> > 
> 
> Hello, @dbateamz I finally solved it with crontab... Is it the correct solution? Have you also used them? so that the virtual machine always boots...
> 
> ```
> /etc/cron.d
> crontab --> @reboot sleep 600 && /usr/local/bin/startubuntuvm.sh
> ```
> 
> file --> startubuntuvm.sh
> 
> ```
> fi
> doas bhyve -c sockets=1,cores=2,threads=2 \
> -m 4G \
> -l bootrom,/usr/local/share/uefi-firmware/BHYVE_BHF_UEFI.fd,/vms/vmubuntu/EFI_VARS.fd \
> -s 0,hostbridge \
> -s 2,fbuf,rfb=0.0.0.0:5900,w=1280,h=1024 \
> -s 3,xhci,tablet \
> -s 10,nvme,/dev/zvol/zroot/vms/vmubuntu/disk0 \
> -s 20,virtio-net,tap0 \
> -s 31,lpc \
> -A -H -P -w \
> vmubuntu
> ```

Hello @dbateamz , 
ctrontab does not solve it, do you know how I can solve it so that the virtual machine starts at startup?...
